### PR TITLE
Pose estimation

### PR DESCRIPTION
**Pull request by garygra**
Estimation package: 
 * cTw: estimate the pose of the robot from aruco tags
 * wTc: states and trajectory from world to camera frame 
 * plant estimator: estimate the current state of the robot from sensing data. Currently, it passes the position from aruco tags.

### CODE REVIEW

File: world/perception/launch/real_time_perc.launch [hunk 1]
```diff
@@ -0,0 +1,13 @@
+<launch>
+	<arg name="control_topic" default="/car/mux/ackermann_cmd_mux/input/navigation" />
```
Comment (aravindsiv) [inline world/perception/launch/real_time_perc.launch]: I thought we are removing the multiplexer and sending directly to the topic?

File: world/perception/include/utils.hpp
Comment (aravindsiv) [inline world/perception/include/utils.hpp:1]: I feel like we have multiple defs of these methods inside the repository (for e.g., in `infrastructure/interface`, `world/mujoco_ros`). What would be the best place to put it so all packages can access it? (Also in favor of creating a `utils` package)

File: infrastructure/interface/msg/marker.msg
Comment (aravindsiv) [inline infrastructure/interface/msg/marker.msg:1]: Nit: Call this `Marker.msg` to stick with ROS conventions

File: infrastructure/interface/include/interface/defs.hpp
Comment (aravindsiv) [inline infrastructure/interface/include/interface/defs.hpp:1]: Previous comment about putting all the functions in `defs.hpp` in one place.

File: infrastructure/interface/include/interface/utils.hpp
Comment (aravindsiv) [inline infrastructure/interface/include/interface/utils.hpp:1]: Previous comment about putting all the functions in `defs.hpp` in one place.

File: estimation/src/aruco_estimation.cpp
Comment (aravindsiv) [inline estimation/src/aruco_estimation.cpp:1]: Is this file needed?

File: estimation/config/cTw_cam0.yaml [hunk 2]
```diff
@@ -0,0 +1,9 @@
+robot_id: 0
+main_marker: 54
+marker_length: 0.167
```
Comment (aravindsiv) [inline estimation/config/cTw_cam0.yaml]: Since markers are square, I would prefer `marker_size` to avoid confusion.

File: world/perception/src/nodelets/aruco_detection.cpp [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+    _header.seq = 0;
+    _header.stamp = ros::Time::now();
+    _header.frame_id = _img_topic_name;
+
+    _markers_msg.header.seq = 0;
+    _markers_msg.header.stamp = ros::Time::now();
+    _markers_msg.header.frame_id = _markers_topic_name;
+
+    _frame_marker_publisher = private_nh.advertise<sensor_msgs::Image>(_img_topic_name, 1);
```
Comment (aravindsiv) [inline world/perception/src/nodelets/aruco_detection.cpp:52]: Why `1`? From [here](https://wiki.ros.org/rospy/Overview/Publishers%20and%20Subscribers#Choosing_a_good_queue_size), it looks like it may not be the worst idea to increase this to say, 10 or more.

File: estimation/launch/mj_estimation.launch [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+	<arg name="simulation" value="true"/>
+
+	<arg name="mujoco_cam_file" value="$(find estimation)/config/mujoco_cam.yaml"/>
+	<arg name="logitech_cam_file" value="$(find estimation)/config/cTw_cam0.yaml"/>
+
+	<arg name="cam_file" if="$(arg simulation)" value="$(arg mujoco_cam_file)" />
+  <arg name="cam_file" unless="$(arg simulation)" value="$(arg logitech_cam_file)" />
+  
+  <group ns="mushr">
```
Comment (aravindsiv) [inline estimation/launch/mj_estimation.launch]: Nit: Add as an argument to this launch file so it can be used across robots.

File: estimation/src/nodelets/aruco_cTw.cpp [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+#include <interface/defs.hpp>
+#include <aruco/aruco_nano.h>
+
+namespace estimation
+{
+class aruco_cTw_nodelet_t : public nodelet::Nodelet
+{
+public:
+  aruco_cTw_nodelet_t()
+    : _robot_pose_topic_name("/aruco_cTw/robot_pose")
```
Comment (aravindsiv) [inline estimation/src/nodelets/aruco_cTw.cpp:30]: Can we default this to the namespace of the robot (e.g. for MuSHR it would be `/mushr`) + `/pose` for simplicity?

File: estimation/src/nodelets/aruco_cTw.cpp [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+      _marker[0].x = marker.x1;
+      _marker[0].y = marker.y1;
+      _marker[1].x = marker.x2;
+      _marker[1].y = marker.y2;
+      _marker[2].x = marker.x3;
+      _marker[2].y = marker.y3;
+      _marker[3].x = marker.x4;
+      _marker[3].y = marker.y4;
+
+      cv::solvePnP(_marker_corners, _marker, _camera_matrix, _dist_coeffs, _Rvec, _Tvec, false, cv::SOLVEPNP_IPPE);
```
Comment (aravindsiv) [inline estimation/src/nodelets/aruco_cTw.cpp:179]: Does this (and some more following lines) need to happen every time `detect()` is called?

File: estimation/launch/mj_estimation.launch [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    <node unless="$(arg simulation)" pkg="nodelet" type="nodelet" name="aruco_detection_nodelet_t" args="load perception/aruco_detection_nodelet_t standalone_nodelet" output="screen">
+        <rosparam command="load" file="$(find perception)/config/aruco_detection.yaml" />
+    </node>
+    <node pkg="nodelet" type="nodelet" name="aruco_cTw_nodelet_t" args="load estimation/aruco_cTw_nodelet_t standalone_nodelet" output="screen">
+      <rosparam command="load" file="$(arg cam_file)" />
+    </node>  
+
+    <node pkg="nodelet" type="nodelet" name="aruco_wTc_nodelet_t" args="load estimation/aruco_wTc_nodelet_t standalone_nodelet" output="screen">
+      <rosparam command="load" file="$(arg cam_file)" />
+      <param name="pose_topic" value="/mushr/aruco_cTw/robot_pose" />
```
Comment (aravindsiv) [inline estimation/launch/mj_estimation.launch]: We should use the `ns` inside the node definitions rather than including it as part of the argument.

File: estimation/launch/mj_estimation.launch [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    <node pkg="nodelet" type="nodelet" name="standalone_nodelet"  args="manager" output="screen"/>
+
+    <node unless="$(arg simulation)" pkg="nodelet" type="nodelet" name="aruco_detection_nodelet_t" args="load perception/aruco_detection_nodelet_t standalone_nodelet" output="screen">
+        <rosparam command="load" file="$(find perception)/config/aruco_detection.yaml" />
+    </node>
+    <node pkg="nodelet" type="nodelet" name="aruco_cTw_nodelet_t" args="load estimation/aruco_cTw_nodelet_t standalone_nodelet" output="screen">
+      <rosparam command="load" file="$(arg cam_file)" />
+    </node>  
+
+    <node pkg="nodelet" type="nodelet" name="aruco_wTc_nodelet_t" args="load estimation/aruco_wTc_nodelet_t standalone_nodelet" output="screen">
```
Comment (aravindsiv) [inline estimation/launch/mj_estimation.launch:23]: Why do we have a world-to-camera frame nodelet?

File: estimation/src/nodelets/plant_estimator.cpp [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+    // const std::vector<prx::system_ptr_t> plants(_plant);
+    // _world_model = std::make_shared<prx::worldmodel_t>(plants, {});
+    _world_model.reset(new prx::world_model_t({ _plant }, {}));
+    _world_model->create_context("estimator_context", { _plant_name }, {});
+    auto context = _world_model->get_context("estimator_context");
+    _system_group = context.first;
+    _collision_group = context.second;
+    _state_space.reset(_system_group->get_state_space());
+    _control_space.reset(_system_group->get_control_space());
+    std::vector<double> min_control_limits = { -1., -1. };
```
Comment (aravindsiv) [inline estimation/src/nodelets/plant_estimator.cpp]: Nit: Should be `-1, -0.5` and `1, 0.5`.

File: estimation/src/nodelets/plant_estimator.cpp
Comment (aravindsiv) [inline estimation/src/nodelets/plant_estimator.cpp:1]: This file is a good addition, but we should discuss how to use it as part of the re-planning.

File: estimation/launch/rosbag_replay.launch
Comment (aravindsiv) [inline estimation/launch/rosbag_replay.launch:1]: What is this file doing?

**Review by aravindsiv — CHANGES_REQUESTED** [review]
Leaving some minor comments on the simulation side (I will continue checking the code for the real-robot perception)

File: estimation/launch/mj_estimation.launch [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+	<arg name="simulation" value="true"/>
+
+	<arg name="mujoco_cam_file" value="$(find estimation)/config/mujoco_cam.yaml"/>
+	<arg name="logitech_cam_file" value="$(find estimation)/config/cTw_cam0.yaml"/>
+
+	<arg name="cam_file" if="$(arg simulation)" value="$(arg mujoco_cam_file)" />
+  <arg name="cam_file" unless="$(arg simulation)" value="$(arg logitech_cam_file)" />
+  
+  <group ns="mushr">
+    <node pkg="nodelet" type="nodelet" name="standalone_nodelet"  args="manager" output="screen"/>
```
Comment (aravindsiv) [inline estimation/launch/mj_estimation.launch:13]: If this file is included from another launch file, this line will throw an error.

File: world/perception/launch/real_time_perc.launch (hunk 1 above)
Comment (garygra) [reply world/perception/launch/real_time_perc.launch]: This is an old launch file. Removing it.

File: world/perception/include/utils.hpp
Comment (garygra) [reply world/perception/include/utils.hpp:1]: Creating `utils` package

File: estimation/src/aruco_estimation.cpp
Comment (garygra) [reply estimation/src/aruco_estimation.cpp:1]: removed

File: world/perception/src/nodelets/aruco_detection.cpp (hunk 3 above)
Comment (garygra) [reply world/perception/src/nodelets/aruco_detection.cpp:52]: Given that this is a sensor, I would prefer to have it as 1. If the network lags, we should focus on the most recent observation rather than old ones.

File: estimation/src/nodelets/aruco_cTw.cpp (hunk 5 above)
Comment (garygra) [reply estimation/src/nodelets/aruco_cTw.cpp:30]: I would prefer not to because this topic should be used by the estimator to do filtering and better estimation of the pose.

File: estimation/src/nodelets/aruco_cTw.cpp (hunk 6 above)
Comment (garygra) [reply estimation/src/nodelets/aruco_cTw.cpp:179]: Some are for viz purposes. Moved to a different function with param to control if is done or not.

File: estimation/launch/mj_estimation.launch (hunk 8 above)
Comment (garygra) [reply estimation/launch/mj_estimation.launch:23]: Primarily for visualization/debugging: check the (estimated/real) poses, trajectories, goal, etc.

File: estimation/src/nodelets/plant_estimator.cpp
Comment (garygra) [reply estimation/src/nodelets/plant_estimator.cpp:1]: Yes. Related to this is my previous comment on cTw: The estimator should be the one that integrates different sensors and does filtering.

File: estimation/launch/rosbag_replay.launch
Comment (garygra) [reply estimation/launch/rosbag_replay.launch:1]: I was using this to check old rosbags. We can remove it if you prefer.

**Review by aravindsiv — APPROVED** [review]
We may need to revisit the following files:
* `rosbag_replay.launch`
* `estimation/src/nodelets/aruco_cTw.cpp`
* `estimation/src/nodelets/aruco_wTc.cpp`

but otherwise looks great. Merging.
